### ISSUE TITLE
Calling `list()` on `urllib.parse.urlparse()` produces a generic type

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#5036]
### Discussed in https://github.com/microsoft/pyright/discussions/5035

<div type='discussions-op-text'>

<sup>Originally posted by **Pyprohly** April 29, 2023</sup>
Before Pyright version 1.1.295, the following code was legal:

```python
import urllib.parse
url = "https://www.google.com?a=1&b=2"
urlparts = list(urllib.parse.urlparse(url))
urlparts[4] = "a=1&b=2&c=3"
url = urllib.parse.urlunparse(urlparts)
```

Pyright (1.1.305) now reports the type of `urlparts` as `list[AnyStr@_ParseResultBase]` instead of `list[str]` as it used to be, making the above code illegal. On the other hand, Mypy (1.2.0) still reports the type as `list[str]`.

I understand that this was caused by [a change in typeshed][1], but the code I’ve written above seems reasonable enough, and it’s not obvious to me what the type of doing `list()` on a generic `NamedTuple` should be.

[1]: https://github.com/python/typeshed/commit/bc847a9b07948a20fcd06ad6c19f2e73437e957e

Is the error reporting for the above code correct? How do I fix the code?

Relevant example:

```python
from typing import NamedTuple, Generic, AnyStr

class _ParseResultBase(NamedTuple, Generic[AnyStr]):
    scheme: AnyStr
    netloc: AnyStr
    path: AnyStr
    params: AnyStr
    query: AnyStr
    fragment: AnyStr

class ParseResult(_ParseResultBase[str]): pass

def urlparse(url: str) -> ParseResult:
    return ParseResult('', '', '', '', '', '')

l = list(urlparse(''))
# Pyright 1.1.294: Type of "l" is "list[str]"
# Pyright 1.1.295+: Type of "l" is "list[AnyStr@_ParseResultBase]"
# Mypy 1.2.0: Revealed type is "builtins.list[builtins.str]"
reveal_type(l)
```
</div>

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.306, which I just published. It will also be included in a future release of pylance.
